docs/language/learn-ql/ql-etudes/river-crossing.ql

```diff
@@ -0,0 +1,114 @@
+/** 
+ * @name River crossing puzzle
+ * @description This implements the classical puzzle where a man is trying to
+ * ferry a goat, a cabbage, and a wolf across a river.
+ * His boat can only take himself and at most one item as cargo.
+ * His problem is that if the goat is left alone with the cabbage,
+ * it will eat it.
+ * And if the wolf is left alone with the goat, it will eat it.
+ * How does he get everything across the river?
+ *
+ * Note: Parts of this QL file are included in the corresponding .rst file.
+ * Make sure to update the line numbers if you change anything here!
+ */
+
+/** A possible cargo item. */
+class Cargo extends string {
+  Cargo() {
+    this = "Nothing" or
+    this = "Goat" or
+    this = "Cabbage" or
+    this = "Wolf"
+  }
+}
+
+/** One of two shores. */
+class Shore extends string {
+  Shore() {
+    this = "Left" or
+    this = "Right"
+  }
+
+  /** Returns the other shore. */
+  Shore other() {
+    this = "Left" and result = "Right"
+    or
+    this = "Right" and result = "Left"
+  }
+}
+
+/** Renders the state as a string. */
+string renderState(Shore man, Shore goat, Shore cabbage, Shore wolf) {
+  result = man + "," + goat + "," + cabbage + "," + wolf
+}
+
+/** A record of where everything is. */
+class State extends string {
```

Comment:
I would prefer each variable declared here to be more obviously linked to a location. Using the cargo name by itself is a tiny bit confusing to me. Would `manLocation`, `goatLocation` etc be clearer? Perhaps that would make later parts of the query look more complicated than is necessary though.... 🤷‍♂

Reply:
Good point. I've changed them all to ``manShore``, ``goatShore`` etc, which makes things clearer in most places. The only problem was the ``eats(predicateShore, preyShore)`` predicate, which looked a bit funny. I've changed it to ``eating``, which possibly helps?

Reply:
I guess you want some synonym of `eatingSituation`, but I can only think of silly ones... (`eatery`, `snackTime`, etc..)